### PR TITLE
crypto/pem: add to standard vlib

### PR DESCRIPTION
**Pull request by phoreverpheebs**
## PEM
This PR adds an implementation of the PEM format based on RFC 1421.

### Functions
Basic `encode` and `decode` functions are included, along with an `encode_config` function for control over the line length and line ending. 

Functions `encode`, `encode_config` and `decode` are covered by tests in `pem_test.v`. The rest of the exposed functions are simple alias functions.

Each public function has a short description of its functionality.

### Enums
Using data from RFC 1421 Section 9 `enum Header` is included to serve as a list of standard PEM headers with a `str()` method for use in writing headers of a new PEM block or indexing headers of a decoded one.

### TL;DR

- [x] Encode
- [x] Decode
- [x] Standard headers
- [x] Documentation

### CODE REVIEW

File: vlib/crypto/pem/decode.v [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	mut rest := data[data.index(pem_begin)?..]
+	mut block := Block{
+		block_type: rest[pem_begin.len..].all_before(pem_eol)
+	}
+	block.headers, rest = parse_headers(rest[pem_begin.len..].all_after(pem_eol).trim_left(' \n\t\v\f\r'))?
+
+	block_end_index := rest.index(pem_end)?
+	b64_data := rest[..block_end_index].replace_each(['\r', '', '\n', '', '\t', '', ' ', ''])
+
+	block.data = []u8{len: block_end_index / 4 * 3, cap: block_end_index / 4 * 3 + 3, init: 0}
```
Comment (spytheman) [inline vlib/crypto/pem/decode.v:18]:
```suggestion
	block_data_len := b64_data.len / 4 * 3
	block.data = []u8{len: block_data_len, cap: block_data_len + 3}
```

File: vlib/crypto/pem/decode.v [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+	return block, rest[rest.index(pem_end)? + pem_end.len..].all_after_first(pem_eol)
+}
+
+fn parse_headers(block string) ?(map[string][]string, string) {
+	headers_str := block.all_before(pem_end).all_before('\n\n')
+
+	// check that something was split or if it's empty
+	if headers_str.len == block.all_before(pem_end).len || headers_str.len == 0 {
+		return map[string][]string{}, block
```
Comment (spytheman) [inline vlib/crypto/pem/decode.v:30]:
```suggestion
		return {}, block
```

File: vlib/crypto/pem/encode.v [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+pub fn (block Block) encode() !string {
+	return block.encode_config(EncodeConfig{})
+}
+
+// `encode_config` encodes the given block into a
+// string using the EncodeConfig. It returns an error if `block_type` is undefined
+// or if a value in `headers` contains an invalid character ':'
+pub fn (block Block) encode_config(config EncodeConfig) !string {
+	if block.block_type == '' {
+		return error('crypto/pem: `encode` called with undefined `block_type`')
```
Comment (spytheman) [inline vlib/crypto/pem/encode.v:17]:
```suggestion
		return error('crypto.pem: `encode` called with undefined `block_type`')
```

File: vlib/crypto/pem/encode.v [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	if block.headers.keys().any(it.contains(':')) || block.headers.values().any(it.contains(':')) {
+		return error('crypto/pem: invalid header character `:`')
```
Comment (spytheman) [inline vlib/crypto/pem/encode.v:20]:
```suggestion
		return error('crypto.pem: invalid header character `:`')
```

File: vlib/crypto/pem/encode.v [hunk 5]
```diff
@@ -0,0 +1,63 @@
+module pem
+
+import encoding.base64
+import arrays
+
+// `encode` encodes `block` into a valid
+// PEM block, with a line length of 64 bytes and line ending of '\r\n'
+pub fn (block Block) encode() !string {
```
Comment (spytheman) [inline vlib/crypto/pem/encode.v:8]: If you add a `[params]` tag over `struct EncodeConfig {`, then you can simply rename encode_config to `encode` and avoid one indirection and you will have 1 public API less.

File: vlib/crypto/pem/encode.v [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	inner += wrap_lines(base64.encode(block.data), newline, length)
+
+	return '${pem_begin}${block.block_type}${pem_eol}${newline}' + '${inner}' +
+		'${pem_end}${block.block_type}${pem_eol}'
+}
+
+[inline]
+fn wrap_lines(str string, newline string, length int) string {
+	return arrays.chunk(str.bytes(), length).map(it.bytestr()).join(newline)
```
Comment (spytheman) [inline vlib/crypto/pem/encode.v:62]: Very elegant.

File: vlib/crypto/pem/pem.v [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	return block.headers[key.str()]
+}
+
+// returns the block data
+//
+// same as `block.data`
+[inline]
+pub fn (block Block) data() []u8 {
+	return block.data
+}
```
Comment (spytheman) [inline vlib/crypto/pem/pem.v:102]: why not just make `.data` a public field?

Comment (spytheman) [reply vlib/crypto/pem/pem.v:102]: Hm, it is already public; then this method can be just deleted.

File: vlib/crypto/pem/pem_test.v [hunk 8]
```diff
@@ -0,0 +1,2496 @@
...
+		}
+		data: [
+			u8(0x2d),
+			0x20,
+			0x41,
+			0x20,
+			0x6d,
+			0x65,
+			0x73,
+			0x73,
```
Comment (spytheman) [inline vlib/crypto/pem/pem_test.v:165]: Adding pairs of `// vfmt off` before, and `// vfmt on` after this block of code, can prevent the wrapping of u8 array elements on single lines.

File: vlib/crypto/pem/pem.v [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+// Headers as described in RFC 1421 Section 9
+pub enum Header {
+	proctype
+	contentdomain
+	dekinfo
+	origid_asymm
+	origid_symm
+	recipid_asymm
+	recipid_symm
```
[the remark this answers is not in the document]
Comment (phoreverpheebs) [reply vlib/crypto/pem/pem.v:28]: i felt like it'd be better to stay true to the naming in the RFC so i stuck to the same (except for those names having a `-` instead of the `_`)

**Comment by phoreverpheebs** [comment]
@spytheman i did all your reviews except for removing the `_`, as personally i'd stick to the definitions in the RFC, but if you think it'd be better to stick to what you said i'll change it

**Comment by spytheman** [comment]
> @spytheman i did all your reviews except for removing the `_`, as personally i'd stick to the definitions in the RFC, but if you think it'd be better to stick to what you said i'll change it

If they are mentioned in the RFC, it is ok, but I am puzzled why they did not use _ for the rest of the enum values as well?

File: vlib/crypto/pem/pem.v (hunk 9 above)
[the remark this answers is not in the document]
Comment (spytheman) [reply vlib/crypto/pem/pem.v:29]: I agree. It should just be consistent - either only using _, or not.

**Comment by phoreverpheebs** [comment]
> > @spytheman i did all your reviews except for removing the `_`, as personally i'd stick to the definitions in the RFC, but if you think it'd be better to stick to what you said i'll change it
> 
> If they are mentioned in the RFC, it is ok, but I am puzzled why they did not use _ for the rest of the enum values as well?

yeah honestly same, but it is an old format, so old decisions i guess. anyway i fixed the [params] messup, should work with the newest commit

File: vlib/crypto/pem/pem.v (hunk 9 above)
[the remark this answers is not in the document]
Comment (spytheman) [reply vlib/crypto/pem/pem.v:29]: For example `Originator-ID-Asymmetric` can be either `originator_id_asymmetric` or `originatoridasymmetric`, but `origid_asymm` seems arbitrary to me.
